### PR TITLE
Adding batch solving capability to the mode solver api 

### PR DESCRIPTION
**Pull request by prashkh**
Made changes to tidy3d.plugis.mode.web.py and tidy3d.web.api.mode.py to add run_batch functionality

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Added new validators in `HeatSimulation`: no structures with dimensions of zero size, no all-Neumann boundary conditions, non-empty simulation domain.
 
+- A batch of Mode Solvers can be run concurrently using `tidy3d.plugins.mode.web.run_batch()`.
```
Comment (tylerflex) [inline CHANGELOG.md:20]: maybe "Mode Solvers" -> "`ModeSolver` objects"

File: tidy3d/web/api/mode.py [hunk 2]
```diff
@@ -137,6 +145,97 @@ def run(
...
+    if results_files is None:
+        results_files = [f"{folder_name}/mode_solver_results_{i}.hdf5" for i in range(len(mode_solvers))]
+
+    results = [None] * len(mode_solvers)
+    
+    # Handling simulation to make sure there are more than one tries if simulation produces some error
+    def handle_simulation(index, retries=0):
+        try:
+            # We'll create a separate progress bar for batch simulations so we'll set verbose=False here for individual runs
+            result = run(mode_solvers[index], f"{task_name}_{index}", f"mode_solver_batch_{index}", folder_name, results_files[index], False, progress_callback_upload, progress_callback_download)
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:209]: could you call this `run` with keyword arguments (eg `task_name=f"{task_name}_{index}"`)? it just helps keep things a bit easier to read and also helps avoids bugs if the arguments change positions later.

File: pyproject.toml
Comment (tylerflex) [inline pyproject.toml:1]: seems like something went a bit awry with the rebase in that some of the changes seem to be applied to 2.6.3. Could you make sure you have the most up to date `pre/2.7` and try again? Ideally there should just be one commit in the PR and none of these extra changes. Sometimes it takes a little trial and error to get right.

File: tidy3d/web/api/mode.py [hunk 3]
```diff
@@ -137,6 +145,97 @@ def run(
...
 
+def run_batch(
+    mode_solvers: List[ModeSolver],
+    task_name: str = "BatchModeSolver",
+    folder_name: str = "BatchModeSolvers",
+    results_files: List[str] = None,
+    verbose: bool = True,
+    max_workers: int = DEFAULT_NUM_WORKERS,
+    max_retries: int = DEFAULT_MAX_RETRIES,
+    retry_delay: int = DEFAULT_RETRY_DELAY,
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:156]: probably this is a `float`?

File: tidy3d/web/api/mode.py [hunk 4]
```diff
@@ -137,6 +145,97 @@ def run(
...
+    progress_callback_upload : Callable[[float], None], optional
+        Optional callback function called when uploading file with ``bytes_in_chunk`` as argument.
+    progress_callback_download : Callable[[float], None], optional
+        Optional callback function called when downloading file with ``bytes_in_chunk`` as argument.
+
+
+    Returns
+    -------
+    List[ModeSolverData]
+        A list of ModeSolverData objects containing the results from each simulation in the batch. None is placed in the list for simulations that fail after all retries.
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:190]: make sure `None` is in double backslash quotes

File: tidy3d/web/api/mode.py [hunk 5]
```diff
@@ -137,6 +145,97 @@ def run(
...
+    """
+    console = get_logging_console()
+
+    # Check type to make sure the user is submitting a list of mode solvers
+    if False in [type(x)== ModeSolver for x in mode_solvers]: 
+        console.print("[red]ERROR: Stopped running a batch of mode solver simulations. Please provide a list of mode solvers of type [cyan]`td.plugins.mode.mode_solver.ModeSolver` [red]to run a batch of mode solver simulations.")
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:196]: I think we're using `console.log` in the rest of the code.

Also, for messages like this going to STDOUT, we want to put the variables in regular single quotes `'`. not backslash quotes. It's a bit confusing..

File: tidy3d/web/api/mode.py [hunk 6]
```diff
@@ -137,6 +145,97 @@ def run(
...
+
+    Returns
+    -------
+    List[ModeSolverData]
+        A list of ModeSolverData objects containing the results from each simulation in the batch. None is placed in the list for simulations that fail after all retries.
+    """
+    console = get_logging_console()
+
+    # Check type to make sure the user is submitting a list of mode solvers
+    if False in [type(x)== ModeSolver for x in mode_solvers]:
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:195]: maybe a bit standard to do
```
if not all(isinstance(x, ModeSolver) for x in mode_solvers):
```

File: tidy3d/web/api/mode.py (hunk 5 above)
Comment (tylerflex) [reply tidy3d/web/api/mode.py:196]: Also probably makes more sense to simply `raise ValueError(...)` here.

File: tidy3d/web/api/mode.py [hunk 7]
```diff
@@ -137,6 +145,97 @@ def run(
...
+        return
+
+    # Result files where the data is stored after downloading
+    if results_files is None:
+        results_files = [f"{folder_name}/mode_solver_results_{i}.hdf5" for i in range(len(mode_solvers))]
+
+    results = [None] * len(mode_solvers)
+    
+    # Handling simulation to make sure there are more than one tries if simulation produces some error
+    def handle_simulation(index, retries=0):
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:206]: can you put the comment as a docstring? 

```py
def handle_simulation(...):
    """Handling simulation to make sure there are more than one tries if simulation produces some error"""
```
and also add type information for the arguments and return?

File: tidy3d/web/api/mode.py [hunk 8]
```diff
@@ -137,6 +145,97 @@ def run(
...
+            if retries < max_retries:
+                # Wait before retrying
+                time.sleep(retry_delay)  
+                handle_simulation(index, retries + 1)
+            else:
+                results[index] = None 
+
+    # Using multithreading to run mode solver simulations concurrently
+    with ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = [executor.submit(handle_simulation, i) for i in range(len(mode_solvers))]
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:221]: `len(mode_solvers)` gets used a lot, maybe just define a `num_mode_solvers = len(mode_solvers)` variable somewhere?

File: tidy3d/web/api/mode.py [hunk 9]
```diff
@@ -137,6 +145,97 @@ def run(
...
+        # Creating a progress bar for batch simulations using rich progress
+        if verbose:
+            with Progress(console=console) as progress:
+                pbar_message = f"[cyan]Running a batch of [deep_pink4]{len(mode_solvers)} mode solvers [cyan]in parallel.\n"
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:225]: some purists might take issue with "in parallel" and could email us for clarification. Maybe we should just leave this part out and just saying that we're running a batch of {N} mode solvers should be enough

File: tidy3d/web/api/mode.py [hunk 10]
```diff
@@ -137,6 +145,97 @@ def run(
...
+            # We'll create a separate progress bar for batch simulations so we'll set verbose=False here for individual runs
+            result = run(mode_solvers[index], f"{task_name}_{index}", f"mode_solver_batch_{index}", folder_name, results_files[index], False, progress_callback_upload, progress_callback_download)
+            results[index] = result
+        except Exception as exc:
+            if retries < max_retries:
+                # Wait before retrying
+                time.sleep(retry_delay)  
+                handle_simulation(index, retries + 1)
+            else:
+                results[index] = None
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:217]: `results[index]` is already initialized as `None` right? so maybe we just `return`? Also, should we `log.warning()` that  the `index`-th mode solver failed?

File: tidy3d/web/api/mode.py [hunk 11]
```diff
@@ -137,6 +145,97 @@ def run(
...
+        if verbose:
+            with Progress(console=console) as progress:
+                pbar_message = f"[cyan]Running a batch of [deep_pink4]{len(mode_solvers)} mode solvers [cyan]in parallel.\n"
+                pbar = progress.add_task(pbar_message, total=len(mode_solvers))
+                for future in as_completed(futures):
+                    progress.advance(pbar)
+                    try:
+                        future.result()  # Add timeout if needed
+                    except TimeoutError:
+                        console.print(f"[red]A ModeSolver task has timed out.")
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:232]: Just wondering, when does this get triggered? eg. if there's a connection error? how long do we wait? Maybe can you explain what this is catching? is it different from the retry code in `handle_simulation`?

Please put `ModeSolver` in single quotes `'` for terminal output. 

Also let's use `console.log()` for consistency

is there any way to get the index of this mode solver in the original list for display?

File: tidy3d/web/api/mode.py (hunk 7 above)
Comment (tylerflex) [reply tidy3d/web/api/mode.py:206]: Should we call this `handle_mode_solver`? not `simulation`? More generally, it seems `simulation` sneaks in the comments and stuff, just checking.

File: tidy3d/web/api/mode.py [hunk 12]
```diff
@@ -137,6 +145,97 @@ def run(
...
+
+                # Make sure the progress bar is complete
+                progress.update(pbar, completed=len(mode_solvers), refresh=True)
+                console.print("[green]Batch simulations completed successfully!")
```
Comment (tylerflex) [inline tidy3d/web/api/mode.py:236]: batch of 'ModeSolver's maybe?

Also, how do we handle this if one of them actually failed? does the progress bar get completed still and do we still print this message?

**Review by tylerflex — CHANGES_REQUESTED** [review]
Thanks @prashkh, this is looking pretty good already. Here are some initial comments to help clean up a few bits. Could you try to resolve them for the next commit with the local mode solve? Thank you!

**Comment by tylerflex** [comment]
@prashkh should we close this in favor of #1654?

**Comment by prashkh** [comment]
@tylerflex Yes, let's close this one and continue it on #1654

**Pull request closed without merging**